src/js_permission_results.hpp

```diff
@@ -0,0 +1,150 @@
+            static void add_listener(ContextType, FunctionType, ObjectType, size_t, const ValueType[], ReturnValue &);
+            static void remove_listener(ContextType, FunctionType, ObjectType, size_t, const ValueType[], ReturnValue &);
+            static void remove_all_listeners(ContextType, FunctionType, ObjectType, size_t, const ValueType[], ReturnValue &);
+
+            std::string const name = "PermissionResults";
+
+            MethodMap<T> const methods = {
+                {"addListener", wrap<add_listener>},
+                {"removeListener", wrap<remove_listener>},
+                {"removeAllListeners", wrap<remove_all_listeners>},
+            };
+
+            PropertyMap<T> const properties = {
+                {"length", {wrap<get_length>, nullptr}},
+            };
+
+            IndexPropertyType<T> const index_accessor = {wrap<get_index>, nullptr};
+        };
+
+        template<typename T>
+        typename T::Object PermissionResultsClass<T>::create_instance(ContextType ctx, PermissionResults<T> *results) {
+            return create_object<T, PermissionResultsClass<T>>(ctx, results);
+        }
+
+        template<typename T>
+        void PermissionResultsClass<T>::get_length(ContextType ctx, ObjectType object, ReturnValue &return_value) {
+            auto results = get_internal<T, PermissionResultsClass<T>>(object);
+            return_value.set((uint32_t)results->size());
+        }
+
+        template<typename T>
+        void PermissionResultsClass<T>::get_index(ContextType ctx, ObjectType object, uint32_t index, ReturnValue &return_value) {
+            auto results = get_internal<T, PermissionResultsClass>(object);
+            auto permission = results->get(index);
+            auto js_permission = Object::create_empty(ctx);
+            Object::set_property(ctx, js_permission, "path", Value::from_string(ctx, permission.path));
+            Object::set_property(ctx, js_permission, "userId", Value::from_string(ctx, permission.condition.user_id));
+
+            if (permission.access == Permission::AccessLevel::Read)
+                Object::set_property(ctx, js_permission, "access", Value::from_string(ctx, "Read"));
+            else if (permission.access == Permission::AccessLevel::Write)
+                Object::set_property(ctx, js_permission, "access", Value::from_string(ctx, "Write"));
+            else if (permission.access == Permission::AccessLevel::Admin)
+                Object::set_property(ctx, js_permission, "access", Value::from_string(ctx, "Admin"));
+            else
+                Object::set_property(ctx, js_permission, "access", Value::from_string(ctx, "None"));
+
+            return_value.set(js_permission);
+        }
+
+        template<typename T>
+        void PermissionResultsClass<T>::add_listener(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType arguments[], ReturnValue &return_value) {
+            validate_argument_count(argc, 1);
```

Comment:
If it's the same to you, let's add a first parameter `eventName` that is ignored (or just asserted on) for now.